### ISSUE TITLE
Fastlane match missing entitlements

### ISSUE THREAD

**Issue by alexlipa91** [fastlane/fastlane#21224]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description
I have a Flutter app that I sign and release through Match.
However when a build with match it seems entitlements get lost (notifications, sign-in, applinks). If I sign and deploy using Xcode UI, using the same profile, the entitlements get populated and work. 

##### Command executed

This is my lane setup:

https://github.com/alexlipa91/nutmeg-flutter/blob/master/ios/fastlane/Fastfile

I pre-build the ipa using `flutter build ipa --release --no-codesign -t lib/screens/Launch.dart`

##### Complete output when running fastlane, including the stack trace and command used

This is the output of the fastlane job https://github.com/alexlipa91/nutmeg-flutter/actions/runs/4758731297/jobs/8457146957

It seems it is using `distribution_profile`, which is the correct one

### Environment

<details>
<details><summary>🚫 fastlane environment 🚫</summary>

### Stack

| Key                         | Value                                                     |
| --------------------------- | --------------------------------------------------------- |
| OS                          | 13.2.1                                                    |
| Ruby                        | 3.1.3                                                     |
| Bundler?                    | false                                                     |
| Git                         | git version 2.31.1                                        |
| Installation Source         | ~/homebrew/Cellar/fastlane/2.211.0_2/libexec/bin/fastlane |
| Host                        | macOS 13.2.1 (22D68)                                      |
| Ruby Lib Dir                | ~/homebrew/Cellar/ruby@3.1/3.1.3_1/lib                    |
| OpenSSL Version             | OpenSSL 3.0.8 7 Feb 2023                                  |
| Is contained                | false                                                     |
| Is homebrew                 | true                                                      |
| Is installed via Fabric.app | false                                                     |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/               |
| Xcode Version               | 14.3                                                      |
| Swift Version               | 5.8                                                       |

### System Locale

| Error                       |
| --------------------------- |
| No Locale with UTF8 found 🚫 |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
default_platform(:ios)

bundle_identifier = "com.nutmeg.app"
app_id = "1592985083"
team_id = "CN7XB83K76"
username = "alessandroliparoti@live.it"

platform :ios do
  lane :closed_beta do

    setup_ci(verbose: true)

    match(
      type: 'appstore',
      app_identifier: bundle_identifier,
      git_basic_authorization: Base64.strict_encode64(ENV["GIT_AUTH_TOKEN"]),
      readonly: true,
      verbose: true,
    )

    gym(
      scheme: "Runner",
      export_method: "app-store",
      export_team_id: team_id,
      skip_build_archive: true,
      archive_path: "../build/ios/archive/Runner.xcarchive",
      verbose: true
    )

    api_key = app_store_connect_api_key(
      key_id: ENV["KEY_ID"],
      issuer_id: "549abdc6-6a6f-40f5-b041-c46daa9d6c91",
      key_filepath: "./AuthKey_" + ENV["KEY_ID"] + ".p8",
    )
    pilot(
      api_key: api_key,
      apple_id: app_id,
      app_identifier: bundle_identifier,
      skip_waiting_for_build_processing: true,
      skip_submission: true,
      distribute_external: false,
      notify_external_testers: false,
      ipa: "./Runner.ipa",
      itc_provider: "CN7XB83K76"
    )
  end

  lane :submit_for_review do |options|
    deliver(
        build_number: options[:build_number],
        app_version: options[:version],
        submit_for_review: true,
        automatic_release: true,
        force: true, # Skip HTMl report verification
        skip_app_version_update: false,
        skip_metadata: false,
        skip_screenshots: true,
        skip_binary_upload: true,
        run_precheck_before_submit: false,
        username: username,
        app_identifier: bundle_identifier,
        submission_information: { add_id_info_uses_idfa: false }
    )
  end
end

```
<details>
<summary>`./fastlane/Appfile`</summary>

```ruby
# app_identifier("com.nutmeg.app") # The bundle identifier of your app
# apple_id("alessandroliparoti@live.it") # Your Apple email address
#
# itc_team_id("123691355") # App Store Connect Team ID
# team_id("CN7XB83K76") # Developer Portal Team ID

# For more information about the Appfile, see:
#     https://docs.fastlane.tools/advanced/#appfile

app_identifier(ENV["DEVELOPER_APP_IDENTIFIER"])
apple_id(ENV["FASTLANE_APPLE_ID"])
itc_team_id(ENV["APP_STORE_CONNECT_TEAM_ID"])
team_id(ENV["DEVELOPER_PORTAL_TEAM_ID"])
```
</details>

### fastlane gems

| Gem      | Version | Update-Status      |
| -------- | ------- | ------------------ |
| fastlane | 2.211.0 | 🚫 Update available |

### Loaded fastlane plugins:

| Plugin                          | Version | Update-Status |
| ------------------------------- | ------- | ------------- |
| fastlane-plugin-flutter_version | 1.1.15  | ✅ Up-To-Date  |

<details><summary><b>Loaded gems</b></summary>

| Gem                             | Version      |
| ------------------------------- | ------------ |
| error_highlight                 | 0.3.0        |
| did_you_mean                    | 1.6.1        |
| atomos                          | 0.1.3        |
| rexml                           | 3.2.5        |
| CFPropertyList                  | 3.0.5        |
| claide                          | 1.1.0        |
| colored2                        | 3.1.2        |
| nanaimo                         | 0.3.0        |
| xcodeproj                       | 1.22.0       |
| rouge                           | 2.0.7        |
| xcpretty                        | 0.3.0        |
| terminal-notifier               | 2.0.0        |
| unicode-display_width           | 1.8.0        |
| terminal-table                  | 1.8.0        |
| plist                           | 3.6.0        |
| public_suffix                   | 5.0.1        |
| addressable                     | 2.8.1        |
| multipart-post                  | 2.0.0        |
| word_wrap                       | 1.0.0        |
| optparse                        | 0.1.1        |
| tty-screen                      | 0.8.1        |
| tty-cursor                      | 0.7.1        |
| tty-spinner                     | 0.9.3        |
| artifactory                     | 3.0.15       |
| babosa                          | 1.0.4        |
| colored                         | 1.2          |
| highline                        | 2.0.3        |
| commander                       | 4.6.0        |
| excon                           | 0.96.0       |
| faraday-em_http                 | 1.0.0        |
| faraday-em_synchrony            | 1.0.0        |
| faraday-excon                   | 1.1.0        |
| faraday-httpclient              | 1.0.1        |
| faraday-multipart               | 1.0.4        |
| faraday-net_http                | 1.0.1        |
| faraday-net_http_persistent     | 1.2.0        |
| faraday-patron                  | 1.0.0        |
| faraday-rack                    | 1.0.0        |
| faraday-retry                   | 1.0.3        |
| ruby2_keywords                  | 0.0.5        |
| faraday                         | 1.10.2       |
| unf_ext                         | 0.0.8.2      |
| unf                             | 0.1.4        |
| domain_name                     | 0.5.20190701 |
| http-cookie                     | 1.0.5        |
| faraday-cookie_jar              | 0.0.7        |
| faraday_middleware              | 1.2.0        |
| fastimage                       | 2.2.6        |
| gh_inspector                    | 1.1.3        |
| json                            | 2.6.1        |
| mini_magick                     | 4.12.0       |
| naturally                       | 2.2.1        |
| rubyzip                         | 2.3.2        |
| security                        | 0.1.3        |
| xcpretty-travis-formatter       | 1.0.1        |
| dotenv                          | 2.8.1        |
| bundler                         | 2.3.26       |
| simctl                          | 1.6.8        |
| jwt                             | 2.6.0        |
| uber                            | 0.1.0        |
| declarative                     | 0.0.20       |
| trailblazer-option              | 0.1.2        |
| representable                   | 3.2.0        |
| retriable                       | 3.1.2        |
| mini_mime                       | 1.1.2        |
| memoist                         | 0.16.2       |
| multi_json                      | 1.15.0       |
| os                              | 1.1.4        |
| signet                          | 0.17.0       |
| googleauth                      | 1.3.0        |
| httpclient                      | 2.8.3        |
| webrick                         | 1.7.0        |
| google-apis-core                | 0.9.4        |
| google-apis-playcustomapp_v1    | 0.12.0       |
| google-apis-androidpublisher_v3 | 0.32.0       |
| google-cloud-env                | 1.6.0        |
| google-cloud-errors             | 1.3.0        |
| google-cloud-core               | 1.6.0        |
| google-apis-iamcredentials_v1   | 0.16.0       |
| google-apis-storage_v1          | 0.19.0       |
| rake                            | 13.0.6       |
| digest-crc                      | 0.6.4        |
| google-cloud-storage            | 1.44.0       |
| emoji_regex                     | 3.2.3        |
| jmespath                        | 1.6.2        |
| aws-partitions                  | 1.689.0      |
| aws-eventstream                 | 1.2.0        |
| aws-sigv4                       | 1.5.2        |
| aws-sdk-core                    | 3.168.4      |
| aws-sdk-kms                     | 1.61.0       |
| aws-sdk-s3                      | 1.117.2      |
| set                             | 1.0.2        |
| forwardable                     | 1.3.2        |
| logger                          | 1.5.0        |
| pathname                        | 0.2.0        |
| shellwords                      | 0.1.0        |
| cgi                             | 0.3.5        |
| date                            | 3.2.2        |
| timeout                         | 0.2.0        |
| stringio                        | 3.0.1        |
| securerandom                    | 0.2.0        |
| uri                             | 0.11.0       |
| openssl                         | 3.0.1        |
| digest                          | 3.1.0        |
| io-nonblock                     | 0.1.0        |
| ipaddr                          | 1.2.4        |
| io-wait                         | 0.2.1        |
| zlib                            | 2.1.1        |
| resolv                          | 0.2.1        |
| time                            | 0.2.0        |
| open-uri                        | 0.2.0        |
| mutex_m                         | 0.1.1        |
| net-http                        | 0.3.0        |
| net-protocol                    | 0.1.2        |
| ostruct                         | 0.5.2        |
| english                         | 0.7.1        |
| erb                             | 2.2.3        |
| strscan                         | 3.0.1        |
| abbrev                          | 0.1.0        |
| io-console                      | 0.5.11       |
| tempfile                        | 0.1.2        |
| delegate                        | 0.2.0        |
| fileutils                       | 1.6.0        |
| tmpdir                          | 0.1.2        |
| base64                          | 0.1.1        |
| singleton                       | 0.1.1        |
| open3                           | 0.1.1        |
| nkf                             | 0.1.1        |
| prettyprint                     | 0.1.1        |
| pp                              | 0.3.0        |
| find                            | 0.1.1        |
| yaml                            | 0.2.0        |
| psych                           | 4.0.4        |
| fastlane-plugin-flutter_version | 1.1.15       |
</details>

*generated on:* **2023-04-21**
</details>

**Comment by nicklbaert** [comment]
@alexlipa91 did you solve this somehow? I ran into the exact same issue :/

**Comment by techouse** [comment]
Check my answer here https://github.com/flutter/flutter/issues/129075#issuecomment-1631063402

**Comment by SPodjasek** [comment]
@techouse Thanks to your comment posted above I've managed to solve this in my case. 

As I use [match](https://docs.fastlane.tools/actions/match/) to manage Apple code signing in Fastlane, so [update_code_signing_settings](https://docs.fastlane.tools/actions/update_code_signing_settings/) wasn't really needed for me.
Root cause of my problems was `skip_build_archive` flag on [gym](https://docs.fastlane.tools/actions/gym/), after disabling it `gym` rebuilds and signs archive with keys and profiles that `match` supplied and uploaded app has proper entitlements in App Store. I should have thought about that in first place, but as a side effect I've read and learned a lot about code signing in Apple ;-). 

FYI this flag was suggested in this [official tutorial](https://docs.flutter.dev/deployment/cd#:~:text=build_app(%0A%20%20skip_build_archive%3A%20true%2C%0A%20%20archive_path%3A%20%22../build/ios/archive/Runner.xcarchive%22%2C%0A))

**Comment by SPodjasek** [comment]
@alexlipa91 I see that you might have the same problem as I've described before. Looking at your code [here](https://github.com/alexlipa91/nutmeg-flutter/blob/1d8086cc3dc30ef2460b9fdbdf56e7487cf8a3f5/ios/fastlane/Fastfile#L36C12-L36C12) I see that you've also disabled archive building in `gym` - try to remove this flag and it should work.

**Comment by kirya355** [comment]
I implemented the solution as described in https://github.com/flutter/flutter/issues/153980. However, there remains a question on how to prevent the archive from being rebuilt. I want to pass `skip_build_archive: true` to `build_app` and ensure that the final build does not lose `Runner.entitlements`.

**Comment by marvin-kolja** [comment]
> I implemented the solution as described in [flutter/flutter#153980](https://github.com/flutter/flutter/issues/153980). However, there remains a question on how to prevent the archive from being rebuilt. I want to pass `skip_build_archive: true` to `build_app` and ensure that the final build does not lose `Runner.entitlements`.

As I understand it, this issue stems from the fact that the `xcarchive` generated by `flutter build ipa --release --no-codesign` is unsigned. Because of this, it does not include the entitlements specified in `Runner/Runner.entitlements` (or any custom entitlement file you've specified). Entitlements are defined through the `CODE_SIGN_ENTITLEMENTS` build setting, which suggests that they are only applied automatically when the app is code-signed during a build action. However, exporting an `xcarchive` to an `.ipa` is not considered a build action (correct me if I'm wrong).

The solution mentioned in https://github.com/fastlane/fastlane/issues/21224#issuecomment-1642139492 works by rebuilding the archive with code signing enabled.

Another viable option I see is resigning the final `.ipa`, for example, using Fastlane’s [resign](https://docs.fastlane.tools/actions/resign/) action. However, it's not that straight forward.

When using `use_app_entitlements: true`, the action applies entitlements from both the provisioning profile and the existing `.ipa`. However, if your app specifies keys in the entitlements file, such as `com.apple.developer.associated-domains`, neither of these sources will have them at this point in time. While you can explicitly provide a custom entitlements file using `entitlements: 'Runner/Runner.entitlements'`, it will only use that file and none of the entitlements from the existing `.ipa`, such as `application-identifier`.

Long story short, a good approach could be to create an entitlements file that includes all required entitlements from different sources such as files, `.ipa`, and provision profiles. Then, resign the `.ipa` using that file. Whether that's a manual thing to do or some custom script doing that on the fly.
